src/workflows/recipes/components/FilterObjectForm.js

```diff
@@ -0,0 +1,803 @@
+//  {"type": "channel", "channels": ["beta"]}
+// We need a mapping from "channel" to "channels", for each, to know what to
+// call it when we serialize this list into a Map where entries look like this:
+//  {"channels": ["beta"]}
+// It is not safe to assume that the "type" can just be the plural form of the word.
+export const MAPPING = {
+  channel: 'channels',
+  country: 'countries',
+  locale: 'locales',
+  version: 'versions',
+};
+
+// In the UI we present the labels but in the data structure we store it by its value.
+// This mapping needs to be ordered to generate the <Select><Option/><Option/>... components.
+// Also, this is used to *display* the human readable version of each type when viewing
+// the details of the filter object.
+export const SAMPLING_TYPES = [
+  { value: 'bucketSample', label: 'Bucket' },
+  { value: 'stableSample', label: 'Stable' },
+  { value: '', label: 'None' },
+];
+
+/* Return a Map from a List.
+  Each item in the list is expected to be something like this:
+
+      [{"type": "age", "ages": [1,2,3]}, ]
+
+  Which is then transformed to this:
+
+      {"ages": [1,2,3]}
+
+  The list of valid keys needs to be known in advance. */
+export const serializeFilterObjectToMap = list => {
+  if (!list) {
+    return Map();
+  }
+  if (!List.isList(list)) {
+    throw new Error(`'${list}' (${typeof list}) is not a List`);
+  }
+
+  return Map(
+    list.map(item => {
+      const type = item.get('type');
+      if (type === 'bucketSample' || type === 'stableSample') {
+        return ['_sampling', item];
+      }
+      const name = MAPPING[type];
+      if (name === undefined) {
+        throw new Error(`Not sure how to convert '${type}'`);
+      }
+      return [name, item.get(name)];
+    }),
+  );
+};
+
+/* Return a List from a Map.
+  Each item in the map is expected to be key'ed based on the reverse of the MAPPING
+  constant. For example, from a list like this:
+
+       {locales: ["sv", "en-US"], ...}
+
+  Transforms to this:
+
+    [{type: "locale", locales: ["sv", "en-US"]}, ...]
+
+  The list of keys have to be known in advance. */
+export const deserializeFilterObjectToList = obj => {
+  if (!obj) {
+    return List();
+  }
+  const reverseMapping = Object.assign(
+    {},
+    ...Object.entries(MAPPING).map(([key, value]) => ({ [value]: key })),
+  );
+  return fromJS(
+    Object.entries(obj)
+      .filter(([key, value]) => {
+        return value && ((key === '_sampling' && value.type) || !!value.length);
+      })
+      .map(([key, value]) => {
+        if (key === '_sampling') {
+          // Special one that needs to be rewritten based on all the keys in the value.
+          // The 'value' should have
+          return Object.assign({}, value);
+        } else {
+          const name = reverseMapping[key];
+          if (name === undefined) {
+            throw new Error(`Key '${key}' no in known mapping.`);
+          }
+          return { [key]: value, type: name };
```

Comment:
The usage here suggests that `reverseMapping` is semantically a mapping, so it should probably be a stdlib Map or Immutable.Map.

Reply:
Yeah, mayhaps. This is kinda annoying. The purpose of this is data here is that it gets converted to a JSON string and sent as part of the HTTP request with `fetch`. I don't see a big win in not keeping it simple.